internal/blockdb/chain.go

```diff
@@ -1,39 +1,43 @@
 package blockdb
 
 import (
-	"bytes"
 	"context"
+	"crypto/sha256"
 	"database/sql"
-	"encoding/json"
 	"fmt"
+
+	"golang.org/x/sync/singleflight"
 )
 
 // Chain tracks its blocks and a block's transactions.
 type Chain struct {
-	db *sql.DB
-	id int64
+	db     *sql.DB
+	id     int64
+	single singleflight.Group
 }
 
-// Txs are transactions expected to be marshalled JSON.
-type Txs [][]byte
+type transactions [][]byte
 
-// TraceBlock tracks a block at height with its transactions.
-// This method is idempotent and can be safely called multiple times with the same arguments.
-func (chain *Chain) TraceBlock(ctx context.Context, height int, txs Txs) error {
-	// TODO(nix 05-27-2022): Presentation in the database layer is generally bad practice. However, the first pass
-	// of this feature requires the user to make raw sql against the database. Therefore, to ease readability
-	// we indent json here. If we have a presentation layer in the future, I suggest removing the json indent here
-	// and let the presentation layer format appropriately.
-	jsonTxs := make([]string, len(txs))
-	buf := new(bytes.Buffer)
-	for i, tx := range txs {
-		if err := json.Indent(buf, tx, "", "  "); err != nil {
-			return fmt.Errorf("block %d: tx %d: malformed json: %w", height, i, err)
-		}
-		jsonTxs[i] = buf.String()
-		buf.Reset()
+func (txs transactions) hash() []byte {
+	h := sha256.New()
+	for _, tx := range txs {
+		h.Write(tx)
 	}
+	return h.Sum(nil)
+}
+
+// SaveBlock tracks a block at height with its transactions.
+// This method is idempotent and can be safely called multiple times with the same arguments.
+// The txs should be human-readable.
+func (chain *Chain) SaveBlock(ctx context.Context, height int, txs [][]byte) error {
+	k := fmt.Sprintf("%d-%x", height, transactions(txs).hash())
+	_, err, _ := chain.single.Do(k, func() (interface{}, error) {
```

Comment:
Single flight maybe overkill now. My original thought was this method may be called a lot. But I kept it because it should work fine.